example/router/router.hpp

```diff
@@ -0,0 +1,260 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/url
+//
+
+#ifndef BOOST_URL_ROUTER_HPP
+#define BOOST_URL_ROUTER_HPP
+
+#include <boost/url/detail/config.hpp>
+#include <detail/router.hpp>
+#include <memory>
+#include <vector>
+#include <stdexcept>
+
+namespace boost {
+namespace urls {
+
+/// Base class for all routers
+class router_base
+{
+    class impl;
+    std::unique_ptr<impl> impl_;
+
+protected:
+    // A node in the resource tree
+    // Each segment in the resource tree might be
+    // associated with
+    struct node
+    {
+        static constexpr std::size_t npos{std::size_t(-1)};
+
+        // literal segment or replacement field
+        detail::segment_template seg{};
+
+        // A pointer to the resource
+        std::size_t resource_idx = std::size_t(-1);
+
+        // The complete match for the resource
+        std::string path_template;
+
+        // Index of the parent node in the
+        // implementation pool of nodes
+        std::size_t parent_idx{npos};
+
+        // Index of child nodes in the pool
+        detail::child_idx_vector child_idx;
+    };
+
+public:
+    class match_results_base
+    {
+    protected:
+        node const* leaf_ = nullptr;
+        string_view* matches_ = nullptr;
+        string_view* ids_ = nullptr;
+        std::size_t size_ = 0;
+
+        match_results_base() = default;
+
+        explicit
+        match_results_base(
+            node const* leaf,
+            string_view* matches,
+            string_view* ids,
+            std::size_t N)
+            : leaf_(leaf)
+            , matches_(matches)
+            , ids_(ids)
+            , size_(N)
+        {}
+    public:
+        using iterator = string_view*;
+        using const_iterator = string_view const*;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+        using reference = string_view&;
+        using const_reference = string_view const&;
+        using pointer = string_view*;
+        using const_pointer = string_view const*;
+
+        // result-like functions
+        explicit
+        operator bool() const;
+
+        bool
+        has_value() const;
+
+        bool
+        has_error() const;
```

Comment:
The matches can have an error?

Reply:
It tried model it like a `result` that already provides the container-related functions when they're valid. The point was to avoid a proxy to a proxy and syntaxes like `v = **m` or `v = (*m)[id]`. It ended up always holding the same type of error though (`grammar::mismatch`) but I could include something else.

Reply:
we have to simplify this, by a lot